File: internal/source/source.go
// Package source defines the release-source abstraction and implements the
// GitHub tag/release source and the custom HTTP source.
package source

import (
	"context"
	"fmt"

	"update/internal/config"
	"update/internal/transport"
)

// Release is the unified release model shared by all sources.
type Release struct {
	Version     string  `json:"version"`
	TagName     string  `json:"tag_name,omitempty"`
	PublishedAt string  `json:"published_at,omitempty"`
	Name        string  `json:"name,omitempty"`
	Notes       string  `json:"notes,omitempty"`
	Checksum    string  `json:"checksum,omitempty"`
	Assets      []Asset `json:"assets"`
}

// Asset is a downloadable file attached to a release.
type Asset struct {
	Name   string `json:"name"`
	URL    string `json:"url"`
	Size   int64  `json:"size,omitempty"`
	SHA256 string `json:"sha256,omitempty"`
}

// Source fetches release information from an upstream.
type Source interface {
	// Latest returns the newest release.
	Latest(ctx context.Context) (*Release, error)
	// List returns releases, newest first, at most limit entries.
	List(ctx context.Context, limit int) ([]*Release, error)
}

// New constructs a Source for the given configuration.
func New(cfg *config.Config, client *transport.Client) (Source, error) {
	switch cfg.Source.Type {
	case "github-tag":
		return &githubSource{cfg: cfg.Source.GitHub, client: client}, nil
	case "custom":
		return &customSource{cfg: cfg.Source.Custom, client: client}, nil
	default:
		return nil, fmt.Errorf("unsupported source type %q", cfg.Source.Type)
	}
}

File: crates/update-core/src/source.rs
// 源抽象与统一发布清单（对应原 Go internal/source/source.go）。
use crate::config::Config;
use crate::custom::CustomSource;
use crate::github::GitHubSource;
use crate::transport::{Client, Error};
use serde::{Deserialize, Serialize};

/// 所有源共享的统一发布模型（docs/design.md §6）。
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Release {
    #[serde(default)]
    pub version: String,
    #[serde(default, rename = "tag_name", skip_serializing_if = "String::is_empty")]
    pub tag_name: String,
    #[serde(
        default,
        rename = "published_at",
        skip_serializing_if = "String::is_empty"
    )]
    pub published_at: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub name: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub notes: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub checksum: String,
    #[serde(default)]
    pub assets: Vec<Asset>,
}

/// 发布所附产物。
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Asset {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub url: String,
    #[serde(default, skip_serializing_if = "is_zero")]
    pub size: i64,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub sha256: String,
}

fn is_zero(n: &i64) -> bool {
    *n == 0
}

/// 发布源。
pub trait Source: Send + Sync {
    /// 最新发布。
    fn latest(&self) -> Result<Release, Error>;
    /// 历史版本，最新在前，最多 limit 条。
    fn list(&self, limit: usize) -> Result<Vec<Release>, Error>;
}

/// 按配置构建源。
pub fn new(cfg: &Config, client: &Client) -> Result<Box<dyn Source>, String> {
    match cfg.source.ty.as_str() {
        "github-tag" => {
            let g = cfg.source.github.clone().ok_or_else(|| {
                "unsupported source: github-tag requires a github section".to_string()
            })?;
            Ok(Box::new(GitHubSource::new(g, client)))
        }
        "custom" => {
            let cu = cfg.source.custom.clone().ok_or_else(|| {
                "unsupported source: custom requires a custom section".to_string()
            })?;
            Ok(Box::new(CustomSource::new(cu, client)))
        }
        other => Err(format!("unsupported source type {other:?}")),
    }
}
